### PR TITLE
propagate connection status

### PR DESCRIPTION
**Pull request by DanielGrossman**

### CODE REVIEW

File: smpperf.go [hunk 1]
```diff
@@ -175,8 +190,10 @@ func (s *SMPPerf) SendMessages() {
 			}
 			time.Sleep(r.Delay())
 			currTransceiver = (currTransceiver + 1) % s.NumSessions
+
+			// here we send the message
 			go func(transceiverIndex int) {
-				sm, err := transceivers[transceiverIndex].Submit(req)
+				sm, err := transceivers[transceiverIndex].transceiver.Submit(req)
```
Comment (cesarrodrig) [inline smpperf.go]: What about here, the class transceiver that you created can do the submission if it has a good connection. If it doesn't, the next transceiver can take the message.

File: smpperf.go [hunk 2]
```diff
@@ -125,23 +130,33 @@ func (s *SMPPerf) SendMessages() {
...
-		go func() {
+		t := &transceiverConn{
+			transceiver: transceiver,
+			err:         nil,
+		}
+
+		// report error on failed conn and Increment error count
+		go func(transceiverIndex int) {
 			for c := range conn {
+				go func() { transceivers[transceiverIndex].err = c.Error() }()
```
Comment (cesarrodrig) [inline smpperf.go]: I think there's a race condition with this statement and the check of `err` below in the code.

File: smpperf.go [hunk 3]
```diff
@@ -175,18 +191,31 @@ func (s *SMPPerf) SendMessages() {
...
+						} else {
+							go sendErrorCount.Increment()
+						}
+					} else {
+						transceiverID := strconv.Itoa(transceiverIndex)
+						msgIDToTransceiverID.Set(sm.RespID(), transceiverID)
+						submittedCount.Increment()
+					}
+				}(currTransceiver)
+				i++
```
Comment (cesarrodrig) [inline smpperf.go]: Why counting the message as successful here and not below `submittedCount.Increment()`?

**Comment by cesarrodrig** [comment]
The counting and handling of transceiver errors looks like it can benefit from some refactor/modularization. I am planning to introduce unit tests and it will be easier if we have everything cleanly separated.

File: smpperf.go (hunk 2 above)
Comment (DanielGrossman) [reply smpperf.go]: would it still be the case if I take it out of the goroutine?

**Comment by DanielGrossman** [comment]
@cesar0094 when you mention counting are you referring to the .Increment() bits, or to the array of transceivers with the error flag?
what should be cleanly separated?

File: smpperf.go [hunk 4]
```diff
@@ -18,18 +19,78 @@ import (
...
+
+func (s *SMPPerf) checkTransceiverErr(transceiverIndex int) bool {
+	s.transceivers[transceiverIndex].m.Lock()
+	defer s.transceivers[transceiverIndex].m.Unlock()
+	if s.transceivers[transceiverIndex].err != nil {
+		return true
+	}
+	return false
+
+}
```
Comment (cesarrodrig) [inline smpperf.go]: `return s.transceivers[transceiverIndex].err != nil`

**Comment by cesarrodrig** [comment]
Besides a small change, looks good to me.

**Pull request merged**